### PR TITLE
C#: Generate strongly-typed method to raise signal events and fix event accessibility

### LINKED ISSUE #66533: Godot 4, C#: Not possible to emit signals with null arguments
### Godot version

v4.0.beta1.mono.official [4ba934bf3]

### System information

Linux

### Issue description

Because the signal arguments of the `EmitSignal` method are now of type `Variant`, and `Variant` is a value type, null signal arguments are no longer possible.

It used to be possible to emit signals with null arguments in Godot 3, and is still possible when using GDScript.

### Steps to reproduce

The following code will not compile.

```csharp
[Signal]
public delegate void ThingHappenedEventHandler(int x, Object obj);

public override void _Ready()
{
    EmitSignal(nameof(ThingHappened), 21, null);
}
```

It results in the compiler error "Cannot convert null to 'Variant' because it is a non-nullable value type" on the line that calls `EmitSignal`.

### Minimal reproduction project

[Godot4CSharpNullSignalArguments.zip](https://github.com/godotengine/godot/files/9664328/Godot4CSharpNullSignalArguments.zip)

### LINKED ISSUE #67549: Godot 4, C#: Not possible to emit signals with enum parameter
### Godot version

4.0b3

### System information

Windows 10

### Issue description

The signal arguments of the EmitSignal method are now of type Variant, and Variant is a value type. But enum types are now no longer possible.

It used to be possible to emit signals with enum arguments in Godot 3.

### Steps to reproduce

Code will not compile anymore (see MRP)
It results in the compiler error "Cannot convert CameraActions to 'Variant'" on the line that calls EmitSignal.

### Minimal reproduction project

[Godot4CSharpEnumSignalArguments.zip](https://github.com/godotengine/godot/files/9803627/Godot4CSharpEnumSignalArguments.zip)

### PR DESCRIPTION
**Pull request by raulsntos**
- Generate signal event with the same accessibility as the delegate
	- Before this, if a user class declares a signal delegate with less accessibility than `public` it resulted in a compiler error because the delegate had less accessibility than the generated event which was always generated as `public`, with this PR the event is generated with the same accessibility that the delegate was declared with.
	- I can't find an issue about this but I'm pretty sure I saw someone affected by this, maybe it was outside of GitHub.
- Generate `On{EventName}` method to raise signal events
	- Closes https://github.com/godotengine/godot/issues/66533
	- Closes https://github.com/godotengine/godot/issues/67549
	- Follows .NET naming conventions but I can change it to `Emit{EventName}` if that's preferred.
	- Method is `protected virtual` unless the class is sealed in which case it's `private`, this follows the .NET event guidelines.
	- Enums are converted to the underlying type so they can be used as Variant in `EmitSignal`.

Sample:

```csharp
// User-defined signal
[Signal]
delegate void MySignalEventHandler(int a, float b, Node c, MyEnum d);

// Method generated by the source generator
protected void OnMySignal(int a, float b, Node c, MyEnum d)
{
    EmitSignal(SignalName.MySignal, a, b, c, (int)d);
}
```

### CODE REVIEW

**Comment by kleonc** [comment]
> Generate `On{EventName}` method to raise signal events
> 
> - Follows .NET naming conventions, (...)

In Godot there's kinda the opposite convention: names of the methods subscribed (connected) to the event (signal) by default start with the `_on_`  prefix (e.g. `_on_Button_pressed`). So `On{EventName}` might be confusing.

> (...) but I can change it to `Emit{EventName}` if that's preferred.

Makes more sense to me.

---

I'd guess you've already considered many different options but do you think it would be possible to avoid incorporating `{EventName}` into the name of the method rasing the event? I mean e.g. having something like a `EmitSignal<EventName>` generic instead of `Emit{EventName}`.

**Comment by raulsntos** [comment]
> In Godot there's kinda the opposite convention: names of the methods subscribed (connected) to the event (signal) by default start with the `_on_` prefix (e.g. `_on_Button_pressed`). So `On{EventName}` might be confusing.

That's true but we already prioritize .NET conventions over Godot conventions in other areas (e.g.: collections). You are probably right though that most users would be more familiar with Godot conventions in this case but I would love to gather more opinions on this before making a decision, that said I don't feel strongly either way about this.

> I'd guess you've already considered many different options but do you think it would be possible to avoid incorporating `{EventName}` into the name of the method rasing the event? I mean e.g. having something like a `EmitSignal<EventName>` generic instead of `Emit{EventName}`.

In C# generic arguments are types and the name of the event is a string so this is not possible. Defining a method for each event follows the .NET guidelines, and since the methods are virtual it also allows you to customize how the event is raised in case the derived class needs it.

Also, keep in mind the `EmitSignal` method still exists if users prefer to use the more dynamic version like you can use `Get` and `Set` instead of the properties. This is necessary because otherwise C# scripts wouldn't be able to interop with more dynamic languages such as GDScript and internally the strongly-typed methods are just a wrapper over the `EmitSignal` method.

**Comment by neikeq** [comment]
There's a conflict between the official .NET convention and the way Godot names callbacks (including the C# docs). I agree it's better to follow the .NET convention for the naming, but we need to think what to do about callbacks.

**Comment by OlliO6** [comment]
I first thought that you could override methods like for example `OnBodyEntered` so you would not need to connect it to yourself anymore, but since the bindings_generator isn't changed, i assume that's not the case. Personally i do agree to @kleonc that renaming `On` to `Emit` would be more understandable, because i think `On` sounds like it is connected to the signal/event but it's not, I mean when you emit the signal via `EmitSignal` it won't be called.

**Comment by raulsntos** [comment]
> but we need to think what to do about callbacks.

Isn't that up to the user to choose? Since they are the ones implementing those methods they are free to call them however they prefer. I guess my preference would be to name those methods after what they do and not after the signal that invokes them.

For example, let's say you have a signal named `PlayerDied` and a callback method that shows the Game Over screen. Instead of calling this method `OnPlayerDied` or `PlayerDiedCallback` I would call it `ShowGameOverScreen`:

```csharp
event Action PlayerDied;

PlayerDied += ShowGameOverScreen;
```

That way the code reads: `When the player dies, show the Game Over screen`.

**Comment by neikeq** [comment]
I meant we should think what to do about callbacks in the Godot documentation. Many examples use the `OnFoo` naming convention. For example, `OnTimerTimeout` in this example: https://docs.godotengine.org/en/latest/getting_started/step_by_step/signals.html#connecting-a-signal-via-code. That one in particular won't cause issues as it includes the signal is in another class, but for connections to self that would be `this.Timeout += OnTimeout` (conflict between `OnTimeout` callback and `OnTimeout` signal emitter).

We can't apply this change before first addressing that.

**Comment by raulsntos** [comment]
In that particular example we could call the method `ToggleVisibility`, this name of each callback would have to be decided on a case-by-case basis. Also, if we are not willing to change GDScript's convention, are we OK with following a different convention for naming callbacks in C#?

**Comment by neikeq** [comment]
> In that particular example we could call the method `ToggleVisibility`, the name of each callback would have to be decided on a case-by-case basis.

That won't work for the connections' dialog, which generates a signal name automatically. Granted, this dialog doesn't generate pascal case names for C# yet, but ideally that should change. Additionally, I don't think it's a good idea to ask doc writers to think about such names. That may discourage those who translate these examples to C#.

We need to come with a generic convention based only on the signal name (and class name if sender != receiver), even if we then recommend using a different convention.

> Also, if we are not willing to change GDScript's convention, are we OK with following a different convention for naming callbacks in C#?

I don't think we have much influence as to change such GDScript conventions. That being said, the signal naming convention in the docs is inconsistent. There are already other examples that use a naming convention like the one you suggest. The reason for needing a generic convention is still there, if only because it's needed for the connections' dialog.

We are perfectly fine with following a different naming convention in C#, that's no problem.

Ideas for a generic callback naming convention: `_SignalName`, `SignalNameCallback`, `SignalNameEmitted` (prepend `ClassName` if sender != receiver).

**Comment by raulsntos** [comment]
To be honest I totally forgot about that dialog but yeah now that you mention it I can see the need for it.

> Ideas for a generic callback naming convention: `_SignalName`, `SignalNameCallback`, `SignalNameEmitted` (prepend `ClassName` if sender != receiver).

I like `SignalNameCallback`.

**Comment by pavlexander** [comment]
Interested to hear updates on this discussion. I can see that the @godotengine/dotnet was assigned as a reviewer - does that mean that the feature is waiting for it's final approval before the release? Or is there more work planned?

**Comment by Reintjuu** [comment]
@raulsntos, is there a way for me to help to get this merged? Do you know what needs to be done?

**Comment by raulsntos** [comment]
This PR needs to be reviewed by the @godotengine/dotnet team. Users can review it too and it's appreciated.

There were previous concerns mentioned about the name of the emit method that conflicts with the name usually given by Godot to signal callbacks (https://github.com/godotengine/godot/pull/68233#issuecomment-1328166410). I suggest to change the callbacks names[^1] to something like `{SignalName}Callback` to avoid the conflict.

The source generator changes can be tested without merging this PR, just copy the source generator code. It won't generate emit methods for native Godot types (e.g.: `Node`) since that needs the changes to `bindings_generator` but it will generate them for user-defined types. Users can also build this PR locally and test with a local NuGet repository to provide feedback.

[^1]: Just to clarify, I mean changing the naming convention for the signal callback name in C#, GDScript would be unaffected.

**Comment by Reintjuu** [comment]
> There were previous concerns mentioned about the name of the emit method that conflicts with the name usually given by Godot to signal callbacks ([#68233 (comment)](https://github.com/godotengine/godot/pull/68233#issuecomment-1328166410)). I suggest to change the callbacks names[1](#user-content-fn-1-88cc207037afd356111a5e23c0465cad) to something like `{SignalName}Callback` to avoid the conflict.

It's about invocation, right? Registering callbacks is up to the user, just like you said in the discussion above, where methods possibly prefixed with `On` are for.

Wouldn't it be better to just name it `Invoke{SignalName}` or even an overload-like naming convention such as `Emit{SignalName}`? Or is that what the naming conflict is about?

**Comment by akien-mga** [comment]
Does this address https://github.com/godotengine/godot-proposals/issues/7891 fully? If so it could be added to the OP.

**Review by paulloz — APPROVED** [review]

**Review by Thomas-X — APPROVED** [review]

**Comment by abelbascu** [comment]
> > but we need to think what to do about callbacks.
> 
> Isn't that up to the user to choose? Since they are the ones implementing those methods they are free to call them however they prefer. I guess my preference would be to name those methods after what they do and not after the signal that invokes them.
> 
> For example, let's say you have a signal named `PlayerDied` and a callback method that shows the Game Over screen. Instead of calling this method `OnPlayerDied` or `PlayerDiedCallback` I would call it `ShowGameOverScreen`:
> 
> ```cs
> event Action PlayerDied;
> 
> PlayerDied += ShowGameOverScreen;
> ```
> 
> That way the code reads: `When the player dies, show the Game Over screen`.

Fully agree with this. As non-expert game dev, there are quite a few use cases where one may want to be more explicit about what the subscribed methods are instead of enforcing to type a unique generic 'SignalNameCallback'.

File: modules/mono/editor/bindings_generator.cpp [hunk 1]
```diff
@@ -3104,6 +3104,46 @@ Error BindingsGenerator::_generate_cs_signal(const BindingsGenerator::TypeInterf
...
+				p_output.append(CLOSE_BLOCK_L1);
+			} else {
+				p_output.append("(");
+
+				StringBuilder cs_emitsignal_params;
+
+				int idx = 0;
+				for (const ArgumentInterface &iarg : p_isignal.arguments) {
+					const TypeInterface *arg_type = _get_type_or_null(iarg.type);
+					ERR_FAIL_NULL_V(arg_type, ERR_BUG); // Argument type not found
```
Comment (AThousandShips) [inline modules/mono/editor/bindings_generator.cpp:3124]: Let's add the improvements to messages here

**Comment by akien-mga** [comment]
Thanks!

**Comment by Delsin-Yu** [comment]
@raulsntos 
From the guidelines you mentioned, shouldn't the generated method be named `Raise{SignalName}` instead of `On{SignalName}`?
The `On{SignalName}` should be used by Callee (the method that fires by the signal) instead of Caller (the method that raises the signal).

**Pull request merged**